openjdk/mmtkRootsClosure.hpp

```diff
@@ -134,5 +134,4 @@ class MMTkScanObjectClosure : public BasicOopIterateClosure {
 //   virtual void do_cld(ClassLoaderData* cld) {
 
 //     printf("CLD: %p", p);
-//   }
-// };
\ No newline at end of file
+// };//   }

```

Comment:
I don't know where the second `// }` is from. Might be a mistake from regex replacement.

Reply:
In that commit, I intended to normalize all `#ifndef`, `#define` and `#endif` directives.  I edited all `.hpp` files with `sed`.  However, it looks like this file `mmtkRootsClosure.hpp` uses CRLF line ending instead of LF, and may have confused `sed`.

I will normalize line endings, too.

By the way, I find that `mmtkRootsClosure.hpp` does not have any `#ifndef` guards. I'll fix that, too.

Reply:
I remember what happened.

1.  That file does not end with any newline characters.
     -   NOTE: Older C/C++ specs require all source files to end with a newline in case some C/C++ preprocessor `#include` a file that does not end with `\n` or `\r\n` will corrupt the next line following the `#include`, but newer C/C++ specs no longer require it.
     -   VSCode, by default, does not insert newline char at the end of a file. But VSCode can be set to do so in the setting.
     -   Vim, Emacs and all traditional text editors automatically insert newline at the end of a file.
2.  My algorithm to match the last `#endif` is to
    1. Reverse all lines with `tac`.
    2. Then use `sed` to match the now "first" `#endif`
    3. Then use `tac` to reverse all lines again.

Because the file did not end with a newline character, the `tac` program got confused, and joined the last two lines in the reverse order.

I think it is worth enforcing UNIX-style line ending (LF) and requiring all files to end with a newline character.